File: apps/ios/dakka/Core/Networking/SocketClient.swift
import Foundation
import Combine
import SocketIO

/// Real-time клиент к бэкенду. Тонкая обёртка над SocketIO Swift,
/// прокидывает события в Combine publishers — Views подписываются.
@MainActor
final class SocketClient: ObservableObject {
    static let shared = SocketClient()

    // MARK: - State

    @Published private(set) var connected: Bool = false

    // MARK: - Event publishers

    let messageReceived  = PassthroughSubject<Message, Never>()
    let messageUpdated   = PassthroughSubject<MessageUpdate, Never>()
    let messageDeleted   = PassthroughSubject<MessageDeleted, Never>()
    let messageRead      = PassthroughSubject<MessageRead, Never>()
    let typingChanged    = PassthroughSubject<TypingEvent, Never>()
    let userStatus       = PassthroughSubject<UserStatusEvent, Never>()
    let chatCreated      = PassthroughSubject<Chat, Never>()
    let chatUpdated      = PassthroughSubject<ChatUpdate, Never>()
    let chatRemoved      = PassthroughSubject<String, Never>()  // chatId
    let chatStateUpdated = PassthroughSubject<ChatStateUpdate, Never>()

    // Calls
    let callIncoming = PassthroughSubject<IncomingCallEvent, Never>()
    let callAccepted = PassthroughSubject<CallAcceptedEvent, Never>()
    let callEnded    = PassthroughSubject<CallEndedEvent, Never>()
    let callSignal   = PassthroughSubject<CallSignalEvent, Never>()

    // MARK: - Internals

    private var manager: SocketManager?
    private var socket: SocketIOClient?

    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.dateDecodingStrategy = .iso8601
        return d
    }()

    private init() {}

    // MARK: - Lifecycle

    /// Подключиться. Идемпотентно — повторный вызов с тем же токеном переподключает.
    func connect() {
        guard let token = KeychainStore.get(KeychainStore.Keys.accessToken) else {
            return
        }

        // Если уже подключены — не дёргаем
        if socket?.status == .connected, manager?.handleQueue == .main { return }

        disconnect()

        let m = SocketManager(socketURL: AppConfig.socketBaseURL, config: [
            .log(false),
            .compress,
            .reconnects(true),
            .reconnectAttempts(-1),
            .reconnectWait(1),
            .forceWebsockets(true),
            .connectParams(["token": token]),
            .extraHeaders(["Authorization": "Bearer \(token)"]),
        ])
        m.handleQueue = .main
        self.manager = m

        let s = m.defaultSocket
        self.socket = s

        wireHandlers(s)
        s.connect()
    }

    func disconnect() {
        socket?.removeAllHandlers()
        socket?.disconnect()
        socket = nil
        manager = nil
        connected = false
    }

    // MARK: - Emitters

    func send(_ payload: SendMessagePayload) {
        emit("message:send", payload)
    }

    func setTyping(chatId: String, isTyping: Bool) {
        socket?.emit("message:typing", ["chatId": chatId, "isTyping": isTyping])
    }

    func markRead(messageId: String, chatId: String) {
        socket?.emit("message:read", ["messageId": messageId, "chatId": chatId])
    }

    func reactToMessage(messageId: String, chatId: String, emoji: String) {
        socket?.emit("message:react", ["messageId": messageId, "chatId": chatId, "emoji": emoji])
    }

    func deleteMessageRemote(messageId: String, chatId: String) {
        socket?.emit("message:delete", ["messageId": messageId, "chatId": chatId])
    }

    func editMessage(messageId: String, chatId: String, content: String, nonce: String? = nil, encrypted: Bool = false) {
        var payload: [String: Any] = [
            "messageId": messageId,
            "chatId": chatId,
            "content": content,
            "encrypted": encrypted,
        ]
        if let n = nonce { payload["nonce"] = n }
        socket?.emit("message:edit", payload)
    }

    // MARK: - Call signaling

    func initiateCall(callId: String, peerId: String, chatId: String, callType: CallType) {
        socket?.emit("call:initiate", [
            "callId": callId,
            "peerId": peerId,
            "chatId": chatId,
            "callType": callType.rawValue,
        ])
    }

    func acceptCall(callId: String) {
        socket?.emit("call:accept", ["callId": callId])
    }

    func rejectCall(callId: String) {
        socket?.emit("call:reject", ["callId": callId])
    }

    func endCall(callId: String) {
        socket?.emit("call:end", ["callId": callId])
    }

    func sendCallSignal(callId: String, signal: CallSignal) {
        guard let data = try? JSONEncoder().encode(signal),
              let obj  = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return }
        socket?.emit("call:signal", ["callId": callId, "signal": obj])
    }

    private func emit<E: Encodable>(_ event: String, _ payload: E) {
        // SendMessagePayload / другие наши Codable всегда сериализуются в dict,
        // явно кастуем чтобы Swift выбрал overload emit(_:with:completion:)
        // принимающий [SocketData] — [String: Any] под него подходит.
        guard let data = try? JSONEncoder().encode(payload),
              let dict = try? JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
        else { return }
        socket?.emit(event, with: [dict], completion: nil)
    }

    // MARK: - Wiring

    private func wireHandlers(_ s: SocketIOClient) {
        s.on(clientEvent: .connect) { [weak self] _, _ in
            self?.connected = true
        }
        s.on(clientEvent: .disconnect) { [weak self] _, _ in
            self?.connected = false
        }
        s.on(clientEvent: .error) { _, data in
            print("[Socket] error:", data)
        }

        s.on("chat:new") { [weak self] data, _ in
            guard let self,
                  let dict = data.first,
                  let chat: Chat = self.decode(dict) else { return }
            self.chatCreated.send(chat)
        }

        s.on("chat:updated") { [weak self] data, _ in
            guard let self,
                  let dict = data.first as? [String: Any],
                  let chatId = dict["chatId"] as? String else { return }
            self.chatUpdated.send(ChatUpdate(
                chatId: chatId,
                name: dict["name"] as? String,
                avatar: dict["avatar"] as? String,
                fields: dict
            ))
        }

        s.on("chat:removed") { [weak self] data, _ in
            guard let self,
                  let dict = data.first as? [String: Any],
                  let chatId = dict["chatId"] as? String else { return }
            self.chatRemoved.send(chatId)
        }

        s.on("chat:state-updated") { [weak self] data, _ in
            guard let self,
                  let dict = data.first as? [String: Any],
                  let chatId = dict["chatId"] as? String else { return }
            let iso = ISO8601DateFormatter()
            iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            let isoBasic = ISO8601DateFormatter()
            // Двухступенчатый парсер: с .withFractionalSeconds и без
            func parseDate(_ key: String) -> Date? {
                guard let raw = dict[key] as? String else { return nil }
                return iso.date(from: raw) ?? isoBasic.date(from: raw)
            }
            let update = ChatStateUpdate(
                chatId: chatId,
                pinned: dict["pinned"] as? Bool,
                archived: dict["archived"] as? Bool,
                pinnedAt: parseDate("pinnedAt"),
                archivedAt: parseDate("archivedAt"),
                mutedUntil: parseDate("mutedUntil"),
                // Бэк может прислать `pinnedAt: null` / `archivedAt: null`, чтобы
                // явно снять флаг. Различаем «ключа нет» и «значение null».
                pinnedAtExplicitlyNull: dict.keys.contains("pinnedAt") && dict["pinnedAt"] is NSNull,
                archivedAtExplicitlyNull: dict.keys.contains("archivedAt") && dict["archivedAt"] is NSNull,
                mutedUntilExplicitlyNull: dict.keys.contains("mutedUntil") && dict["mutedUntil"] is NSNull
            )
            self.chatStateUpdated.send(update)
        }

        s.on("chat:member-left") { [weak self] data, _ in
            guard let self,
                  let dict = data.first as? [String: Any],
                  let chatId = dict["chatId"] as? String else { return }
            // Просто посылаем chat:updated — UI рефрешит
            self.chatUpdated.send(ChatUpdate(chatId: chatId, name: nil, avatar: nil, fields: dict))
        }

        s.on("message:new") { [weak self] data, _ in
            guard let self,
                  let dict = data.first,
                  let msg: Message = self.decode(dict) else { return }
            self.messageReceived.send(msg)
        }

        s.on("message:updated") { [weak self] data, _ in
            guard let self,
                  let dict = data.first as? [String: Any],
                  let id = dict["id"] as? String else { return }
            self.messageUpdated.send(MessageUpdate(id: id, fields: dict))
        }

        s.on("message:deleted") { [weak self] data, _ in
            guard let self,
                  let dict = data.first as? [String: Any],
                  let messageId = dict["messageId"] as? String,
                  let chatId    = dict["chatId"] as? String else { return }
            self.messageDeleted.send(.init(messageId: messageId, chatId: chatId))
        }

        s.on("message:read") { [weak self] data, _ in
            guard let self,
                  let dict = data.first as? [String: Any],
                  let messageId = dict["messageId"] as? String,
                  let chatId    = dict["chatId"]    as? String,
                  let userId    = dict["userId"]    as? String else { return }
            let readAt = (dict["readAt"] as? String).flatMap { ISO8601DateFormatter().date(from: $0) } ?? Date()
            self.messageRead.send(.init(messageId: messageId, chatId: chatId, userId: userId, readAt: readAt))
        }

        s.on("user:typing") { [weak self] data, _ in
            guard let self,
                  let dict = data.first as? [String: Any],
                  let chatId  = dict["chatId"]  as? String,
                  let userId  = dict["userId"]  as? String,
                  let typing  = dict["isTyping"] as? Bool else { return }
            self.typingChanged.send(.init(chatId: chatId, userId: userId, isTyping: typing))
        }

        s.on("user:status") { [weak self] data, _ in
            guard let self,
                  let dict = data.first as? [String: Any],
                  let userId = dict["userId"] as? String,
                  let status = dict["status"] as? String else { return }
            self.userStatus.send(.init(userId: userId, status: status))
        }

        // ── Calls ──
        s.on("call:incoming") { [weak self] data, _ in
            guard let self,
                  let dict = data.first as? [String: Any],
                  let callId   = dict["callId"]   as? String,
                  let chatId   = dict["chatId"]   as? String,
                  let callerId = dict["callerId"] as? String,
                  let typeRaw  = dict["callType"] as? String,
                  let type     = CallType(rawValue: typeRaw) else { return }
            let event = IncomingCallEvent(
                callId: callId,
                chatId: chatId,
                callerId: callerId,
                callerName: (dict["callerName"] as? String) ?? "Звонок",
                callerAvatar: dict["callerAvatar"] as? String,
                callType: type
            )
            self.callIncoming.send(event)
        }

        s.on("call:accepted") { [weak self] data, _ in
            guard let self,
                  let dict = data.first as? [String: Any],
                  let callId = dict["callId"] as? String,
                  let peerId = dict["peerId"] as? String else { return }
            self.callAccepted.send(.init(callId: callId, peerId: peerId))
        }

        s.on("call:ended") { [weak self] data, _ in
            guard let self,
                  let dict = data.first as? [String: Any],
                  let callId = dict["callId"] as? String else { return }
            self.callEnded.send(.init(callId: callId, reason: dict["reason"] as? String))
        }

        s.on("call:signal") { [weak self] data, _ in
            guard let self,
                  let dict = data.first as? [String: Any],
                  let callId = dict["callId"] as? String,
                  let sigDict = dict["signal"] as? [String: Any] else { return }
            // Декодируем CallSignal через JSON-round-trip
            guard let raw = try? JSONSerialization.data(withJSONObject: sigDict),
                  let signal = try? JSONDecoder().decode(CallSignal.self, from: raw)
            else { return }
            self.callSignal.send(.init(callId: callId, signal: signal))
        }
    }

    /// Decoder для произвольного JSON-объекта (socket.io отдаёт `[Any]`).
    private func decode<T: Decodable>(_ raw: Any) -> T? {
        guard let data = try? JSONSerialization.data(withJSONObject: raw)
        else { return nil }
        return try? decoder.decode(T.self, from: data)
    }
}

// MARK: - Event payloads

struct ChatUpdate {
    let chatId: String
    let name: String?
    let avatar: String?
    let fields: [String: Any]
}

/// Событие `chat:state-updated`.
///
/// Бэк присылает либо булевый флаг (`pinned: true`), либо точное время
/// (`pinnedAt: "2026-06-01T…"` / `null`). Здесь храним оба, плюс маркеры
/// «значение прислали как `null`» — чтобы UI мог честно снять флаг.
struct ChatStateUpdate {
    let chatId: String
    let pinned: Bool?
    let archived: Bool?
    let pinnedAt: Date?
    let archivedAt: Date?
    let mutedUntil: Date?
    let pinnedAtExplicitlyNull: Bool
    let archivedAtExplicitlyNull: Bool
    let mutedUntilExplicitlyNull: Bool
}

struct MessageUpdate {
    let id: String
    let fields: [String: Any]
}

struct MessageDeleted {
    let messageId: String
    let chatId: String
}

struct MessageRead {
    let messageId: String
    let chatId: String
    let userId: String
    let readAt: Date
}

struct TypingEvent {
    let chatId: String
    let userId: String
    let isTyping: Bool
}

struct UserStatusEvent {
    let userId: String
    let status: String
}

File: apps/android/app/src/main/java/online/akkdmsg/dakka/data/api/SocketClient.kt
package online.akkdmsg.dakka.data.api

import io.socket.client.IO
import io.socket.client.Socket
import io.socket.engineio.client.transports.WebSocket
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.json.Json
import online.akkdmsg.dakka.call.CallAcceptedEvent
import online.akkdmsg.dakka.call.CallEndedEvent
import online.akkdmsg.dakka.call.CallSignal
import online.akkdmsg.dakka.call.CallSignalEvent
import online.akkdmsg.dakka.call.CallType
import online.akkdmsg.dakka.call.IncomingCallEvent
import online.akkdmsg.dakka.data.AppConfig
import online.akkdmsg.dakka.data.Chat
import online.akkdmsg.dakka.data.Message
import online.akkdmsg.dakka.data.SendMessagePayload
import org.json.JSONObject
import java.net.URI

/**
 * Real-time клиент к бэкенду через socket.io v4.
 * Один синглтон, подключается при логине, отписывается при logout.
 *
 * Compose-стороны подписываются на Flow-публикаторы через collectAsState/LaunchedEffect.
 */
object SocketClient {

    // ── State ────────────────────────────────────────────────────────────────

    private val _connected = MutableStateFlow(false)
    val connected: StateFlow<Boolean> = _connected.asStateFlow()

    // ── Event flows ──────────────────────────────────────────────────────────

    private val _messageReceived = MutableSharedFlow<Message>(extraBufferCapacity = 32)
    val messageReceived: SharedFlow<Message> = _messageReceived.asSharedFlow()

    private val _messageDeleted = MutableSharedFlow<MessageDeleted>(extraBufferCapacity = 16)
    val messageDeleted: SharedFlow<MessageDeleted> = _messageDeleted.asSharedFlow()

    private val _messageUpdated = MutableSharedFlow<String>(extraBufferCapacity = 16) // messageId
    val messageUpdated: SharedFlow<String> = _messageUpdated.asSharedFlow()

    private val _messageRead = MutableSharedFlow<MessageRead>(extraBufferCapacity = 16)
    val messageRead: SharedFlow<MessageRead> = _messageRead.asSharedFlow()

    private val _typingChanged = MutableSharedFlow<TypingEvent>(extraBufferCapacity = 16)
    val typingChanged: SharedFlow<TypingEvent> = _typingChanged.asSharedFlow()

    private val _userStatus = MutableSharedFlow<UserStatusEvent>(extraBufferCapacity = 16)
    val userStatus: SharedFlow<UserStatusEvent> = _userStatus.asSharedFlow()

    private val _chatCreated = MutableSharedFlow<Chat>(extraBufferCapacity = 8)
    val chatCreated: SharedFlow<Chat> = _chatCreated.asSharedFlow()

    private val _chatUpdated = MutableSharedFlow<String>(extraBufferCapacity = 8) // chatId
    val chatUpdated: SharedFlow<String> = _chatUpdated.asSharedFlow()

    private val _chatRemoved = MutableSharedFlow<String>(extraBufferCapacity = 8) // chatId
    val chatRemoved: SharedFlow<String> = _chatRemoved.asSharedFlow()

    private val _chatStateUpdated = MutableSharedFlow<ChatStateUpdate>(extraBufferCapacity = 16)
    val chatStateUpdated: SharedFlow<ChatStateUpdate> = _chatStateUpdated.asSharedFlow()

    // Calls
    private val _callIncoming = MutableSharedFlow<IncomingCallEvent>(extraBufferCapacity = 4)
    val callIncoming: SharedFlow<IncomingCallEvent> = _callIncoming.asSharedFlow()
    private val _callAccepted = MutableSharedFlow<CallAcceptedEvent>(extraBufferCapacity = 4)
    val callAccepted: SharedFlow<CallAcceptedEvent> = _callAccepted.asSharedFlow()
    private val _callEnded = MutableSharedFlow<CallEndedEvent>(extraBufferCapacity = 4)
    val callEnded: SharedFlow<CallEndedEvent> = _callEnded.asSharedFlow()
    private val _callSignal = MutableSharedFlow<CallSignalEvent>(extraBufferCapacity = 64)
    val callSignal: SharedFlow<CallSignalEvent> = _callSignal.asSharedFlow()

    // ── Internal ─────────────────────────────────────────────────────────────

    private var socket: Socket? = null
    private val json = Json { ignoreUnknownKeys = true; coerceInputValues = true }

    // ── Lifecycle ────────────────────────────────────────────────────────────

    /** Подключиться. Идемпотентно — повторный вызов переподключает. */
    fun connect() {
        val token = ApiClient.accessToken ?: return
        if (socket?.connected() == true) return

        disconnect()

        val opts = IO.Options().apply {
            forceNew = true
            reconnection = true
            reconnectionDelay = 1000
            reconnectionAttempts = Int.MAX_VALUE
            transports = arrayOf(WebSocket.NAME)
            auth = mutableMapOf("token" to token)
            extraHeaders = mapOf("Authorization" to listOf("Bearer $token"))
        }

        val s = IO.socket(URI.create(AppConfig.SOCKET_BASE_URL), opts)
        socket = s
        wireHandlers(s)
        s.connect()
    }

    fun disconnect() {
        socket?.off()
        socket?.disconnect()
        socket = null
        _connected.value = false
    }

    // ── Emitters ─────────────────────────────────────────────────────────────

    fun send(payload: SendMessagePayload) {
        val jsonStr = json.encodeToString(SendMessagePayload.serializer(), payload)
        socket?.emit("message:send", JSONObject(jsonStr))
    }

    fun setTyping(chatId: String, isTyping: Boolean) {
        socket?.emit("message:typing", JSONObject().apply {
            put("chatId", chatId); put("isTyping", isTyping)
        })
    }

    fun markRead(messageId: String, chatId: String) {
        socket?.emit("message:read", JSONObject().apply {
            put("messageId", messageId); put("chatId", chatId)
        })
    }

    fun reactToMessage(messageId: String, chatId: String, emoji: String) {
        socket?.emit("message:react", JSONObject().apply {
            put("messageId", messageId); put("chatId", chatId); put("emoji", emoji)
        })
    }

    fun deleteMessageRemote(messageId: String, chatId: String) {
        socket?.emit("message:delete", JSONObject().apply {
            put("messageId", messageId); put("chatId", chatId)
        })
    }

    fun editMessage(messageId: String, chatId: String, content: String, nonce: String? = null, encrypted: Boolean = false) {
        socket?.emit("message:edit", JSONObject().apply {
            put("messageId", messageId)
            put("chatId", chatId)
            put("content", content)
            put("encrypted", encrypted)
            if (nonce != null) put("nonce", nonce)
        })
    }

    // ── Call signaling ───────────────────────────────────────────────────────

    fun initiateCall(callId: String, peerId: String, chatId: String, callType: CallType) {
        socket?.emit("call:initiate", JSONObject().apply {
            put("callId", callId); put("peerId", peerId); put("chatId", chatId); put("callType", callType.raw)
        })
    }
    fun acceptCall(callId: String)  { socket?.emit("call:accept", JSONObject().put("callId", callId)) }
    fun rejectCall(callId: String)  { socket?.emit("call:reject", JSONObject().put("callId", callId)) }
    fun endCall(callId: String)     { socket?.emit("call:end",    JSONObject().put("callId", callId)) }

    fun sendCallSignal(callId: String, signal: CallSignal) {
        val signalJson = json.encodeToString(CallSignal.serializer(), signal)
        socket?.emit("call:signal", JSONObject().apply {
            put("callId", callId)
            put("signal", JSONObject(signalJson))
        })
    }

    // ── Wiring ───────────────────────────────────────────────────────────────

    private fun wireHandlers(s: Socket) {
        s.on(Socket.EVENT_CONNECT)    { _connected.value = true }
        s.on(Socket.EVENT_DISCONNECT) { _connected.value = false }

        s.on("message:new") { args ->
            decodeJson(args.firstOrNull(), Message.serializer())?.let { _messageReceived.tryEmit(it) }
        }

        s.on("message:deleted") { args ->
            val o = args.firstOrNull() as? JSONObject ?: return@on
            val msgId  = o.optString("messageId").takeIf { it.isNotEmpty() } ?: return@on
            val chatId = o.optString("chatId").takeIf { it.isNotEmpty() } ?: return@on
            _messageDeleted.tryEmit(MessageDeleted(msgId, chatId))
        }

        s.on("message:updated") { args ->
            val o = args.firstOrNull() as? JSONObject ?: return@on
            val id = o.optString("id").takeIf { it.isNotEmpty() } ?: return@on
            _messageUpdated.tryEmit(id)
        }

        s.on("message:read") { args ->
            val o = args.firstOrNull() as? JSONObject ?: return@on
            _messageRead.tryEmit(
                MessageRead(
                    messageId = o.optString("messageId"),
                    chatId    = o.optString("chatId"),
                    userId    = o.optString("userId"),
                    readAt    = o.optString("readAt"),
                )
            )
        }

        s.on("user:typing") { args ->
            val o = args.firstOrNull() as? JSONObject ?: return@on
            _typingChanged.tryEmit(
                TypingEvent(
                    chatId   = o.optString("chatId"),
                    userId   = o.optString("userId"),
                    isTyping = o.optBoolean("isTyping"),
                )
            )
        }

        s.on("user:status") { args ->
            val o = args.firstOrNull() as? JSONObject ?: return@on
            _userStatus.tryEmit(
                UserStatusEvent(
                    userId = o.optString("userId"),
                    status = o.optString("status"),
                )
            )
        }

        s.on("chat:new") { args ->
            decodeJson(args.firstOrNull(), Chat.serializer())?.let { _chatCreated.tryEmit(it) }
        }

        s.on("chat:updated") { args ->
            val o = args.firstOrNull() as? JSONObject ?: return@on
            o.optString("chatId").takeIf { it.isNotEmpty() }?.let { _chatUpdated.tryEmit(it) }
        }

        s.on("chat:removed") { args ->
            val o = args.firstOrNull() as? JSONObject ?: return@on
            o.optString("chatId").takeIf { it.isNotEmpty() }?.let { _chatRemoved.tryEmit(it) }
        }

        s.on("chat:member-left") { args ->
            val o = args.firstOrNull() as? JSONObject ?: return@on
            o.optString("chatId").takeIf { it.isNotEmpty() }?.let { _chatUpdated.tryEmit(it) }
        }

        s.on("chat:state-updated") { args ->
            val o = args.firstOrNull() as? JSONObject ?: return@on
            val chatId = o.optString("chatId").takeIf { it.isNotEmpty() } ?: return@on
            // Различаем "поле отсутствует" (нет изменений) vs "поле = null" (сбросить).
            // Для строки: hasKey=false → null + present=false; hasKey=true, null → null + present=true.
            fun str(key: String): Pair<String?, Boolean> =
                if (!o.has(key)) null to false
                else if (o.isNull(key)) null to true
                else o.optString(key).takeIf { it.isNotEmpty() } to true
            fun bool(key: String): Pair<Boolean?, Boolean> =
                if (!o.has(key)) null to false
                else if (o.isNull(key)) null to true
                else o.optBoolean(key) to true
            val pinned = bool("pinned").first
            val archived = bool("archived").first
            val pinnedAtPair = str("pinnedAt")
            val archivedAtPair = str("archivedAt")
            val mutedUntilPair = str("mutedUntil")
            _chatStateUpdated.tryEmit(
                ChatStateUpdate(
                    chatId = chatId,
                    pinned = pinned,
                    archived = archived,
                    pinnedAt = pinnedAtPair.first,
                    archivedAt = archivedAtPair.first,
                    mutedUntil = mutedUntilPair.first,
                    pinnedAtPresent = pinnedAtPair.second,
                    archivedAtPresent = archivedAtPair.second,
                    mutedUntilPresent = mutedUntilPair.second,
                )
            )
        }

        // ── Calls ──
        s.on("call:incoming") { args ->
            val o = args.firstOrNull() as? JSONObject ?: return@on
            _callIncoming.tryEmit(IncomingCallEvent(
                callId       = o.optString("callId"),
                chatId       = o.optString("chatId"),
                callerId     = o.optString("callerId"),
                callerName   = o.optString("callerName", "Звонок"),
                callerAvatar = o.optString("callerAvatar").takeIf { it.isNotEmpty() },
                callType     = CallType.from(o.optString("callType")),
            ))
        }
        s.on("call:accepted") { args ->
            val o = args.firstOrNull() as? JSONObject ?: return@on
            _callAccepted.tryEmit(CallAcceptedEvent(
                callId = o.optString("callId"),
                peerId = o.optString("peerId"),
            ))
        }
        s.on("call:ended") { args ->
            val o = args.firstOrNull() as? JSONObject ?: return@on
            _callEnded.tryEmit(CallEndedEvent(
                callId = o.optString("callId"),
                reason = o.optString("reason").takeIf { it.isNotEmpty() },
            ))
        }
        s.on("call:signal") { args ->
            val o = args.firstOrNull() as? JSONObject ?: return@on
            val callId = o.optString("callId").takeIf { it.isNotEmpty() } ?: return@on
            val sig = o.optJSONObject("signal") ?: return@on
            val signal = runCatching {
                json.decodeFromString(CallSignal.serializer(), sig.toString())
            }.getOrNull() ?: return@on
            _callSignal.tryEmit(CallSignalEvent(callId, signal))
        }
    }

    /** Декодирует org.json.JSONObject в @Serializable через JSON-roundtrip. */
    private fun <T> decodeJson(raw: Any?, serializer: kotlinx.serialization.KSerializer<T>): T? {
        val str = raw?.toString() ?: return null
        return runCatching { json.decodeFromString(serializer, str) }.getOrNull()
    }
}

// ── Event DTOs ───────────────────────────────────────────────────────────────

data class MessageDeleted(val messageId: String, val chatId: String)
data class MessageRead(val messageId: String, val chatId: String, val userId: String, val readAt: String)
data class TypingEvent(val chatId: String, val userId: String, val isTyping: Boolean)
data class UserStatusEvent(val userId: String, val status: String)

/**
 * Событие 'chat:state-updated' — изменения pin/archive/mute текущего пользователя.
 * `*Present` различает "поле не пришло — оставить как есть" от "поле = null — сбросить".
 */
data class ChatStateUpdate(
    val chatId: String,
    val pinned: Boolean?,
    val archived: Boolean?,
    val pinnedAt: String?,
    val archivedAt: String?,
    val mutedUntil: String?,
    val pinnedAtPresent: Boolean = true,
    val archivedAtPresent: Boolean = true,
    val mutedUntilPresent: Boolean = true,
)
